File: src/builtins/commandline.cpp
// Functions used for implementing the commandline builtin.
#include "config.h"  // IWYU pragma: keep

#include "commandline.h"

#include <algorithm>
#include <cerrno>
#include <cwchar>
#include <string>

#include "../builtin.h"
#include "../common.h"
#include "../fallback.h"  // IWYU pragma: keep
#include "../input.h"
#include "../input_common.h"
#include "../io.h"
#include "../maybe.h"
#include "../parse_constants.h"
#include "../parse_util.h"
#include "../parser.h"
#include "../proc.h"
#include "../reader.h"
#include "../tokenizer.h"
#include "../wgetopt.h"
#include "../wutil.h"  // IWYU pragma: keep
#include "builtins/shared.rs.h"

/// Which part of the comandbuffer are we operating on.
enum {
    STRING_MODE = 1,  // operate on entire buffer
    JOB_MODE,         // operate on job under cursor
    PROCESS_MODE,     // operate on process under cursor
    TOKEN_MODE        // operate on token under cursor
};

/// For text insertion, how should it be done.
enum {
    REPLACE_MODE = 1,  // replace current text
    INSERT_MODE,       // insert at cursor position
    APPEND_MODE        // insert at end of current token/command/buffer
};

/// Handle a single readline_cmd_t command out-of-band.
void reader_handle_command(readline_cmd_t cmd);

/// Replace/append/insert the selection with/at/after the specified string.
///
/// \param begin beginning of selection
/// \param end end of selection
/// \param insert the string to insert
/// \param append_mode can be one of REPLACE_MODE, INSERT_MODE or APPEND_MODE, affects the way the
/// test update is performed
/// \param buff the original command line buffer
/// \param cursor_pos the position of the cursor in the command line
static void replace_part(const wchar_t *begin, const wchar_t *end, const wchar_t *insert,
                         int append_mode, const wchar_t *buff, size_t cursor_pos) {
    size_t out_pos = cursor_pos;

    wcstring out;

    out.append(buff, begin - buff);

    switch (append_mode) {
        case REPLACE_MODE: {
            out.append(insert);
            out_pos = out.size();
            break;
        }
        case APPEND_MODE: {
            out.append(begin, end - begin);
            out.append(insert);
            break;
        }
        case INSERT_MODE: {
            long cursor = cursor_pos - (begin - buff);
            out.append(begin, cursor);
            out.append(insert);
            out.append(begin + cursor, end - begin - cursor);
            out_pos += std::wcslen(insert);
            break;
        }
        default: {
            DIE("unexpected append_mode");
        }
    }
    out.append(end);
    commandline_set_buffer(out, out_pos);
}

/// Output the specified selection.
///
/// \param begin start of selection
/// \param end  end of selection
/// \param cut_at_cursor whether printing should stop at the surrent cursor position
/// \param tokenize whether the string should be tokenized, printing one string token on every line
/// and skipping non-string tokens
/// \param buffer the original command line buffer
/// \param cursor_pos the position of the cursor in the command line
static void write_part(const wchar_t *begin, const wchar_t *end, int cut_at_cursor, int tokenize,
                       const wchar_t *buffer, size_t cursor_pos, io_streams_t &streams) {
    size_t pos = cursor_pos - (begin - buffer);

    if (tokenize) {
        // std::fwprintf( stderr, L"Subshell: %ls, end char %lc\n", buff, *end );
        wcstring out;
        wcstring buff(begin, end - begin);
        auto tok = new_tokenizer(buff.c_str(), TOK_ACCEPT_UNFINISHED);
        while (auto token = tok->next()) {
            if ((cut_at_cursor) && (token->offset + token->length >= pos)) break;

            if (token->type_ == token_type_t::string) {
                wcstring tmp = *tok->text_of(*token);
                auto maybe_unescaped = unescape_string(tmp.c_str(), tmp.size(), UNESCAPE_INCOMPLETE,
                                                       STRING_STYLE_SCRIPT);
                assert(maybe_unescaped);
                out.append(*maybe_unescaped);
                out.push_back(L'\n');
            }
        }

        streams.out()->append(out);
    } else {
        if (cut_at_cursor) {
            streams.out()->append(wcstring{begin, pos});
        } else {
            streams.out()->append(wcstring{begin, end});
        }
        streams.out()->push(L'\n');
    }
}

/// The commandline builtin. It is used for specifying a new value for the commandline.
int builtin_commandline(const void *_parser, void *_streams, void *_argv) {
    const auto &parser = *static_cast<const parser_t *>(_parser);
    auto &streams = *static_cast<io_streams_t *>(_streams);
    auto argv = static_cast<const wchar_t **>(_argv);
    const commandline_state_t rstate = commandline_get_state();
    const wchar_t *cmd = argv[0];
    int buffer_part = 0;
    bool cut_at_cursor = false;

    int argc = builtin_count_args(argv);
    int append_mode = 0;

    bool function_mode = false;
    bool selection_mode = false;

    bool tokenize = false;

    bool cursor_mode = false;
    bool selection_start_mode = false;
    bool selection_end_mode = false;
    bool line_mode = false;
    bool search_mode = false;
    bool paging_mode = false;
    bool paging_full_mode = false;
    bool is_valid = false;
    const wchar_t *begin = nullptr, *end = nullptr;
    const wchar_t *override_buffer = nullptr;

    const auto &ld = parser.libdata();

    static const wchar_t *const short_options = L":abijpctforhI:CBELSsP";
    static const struct woption long_options[] = {{L"append", no_argument, 'a'},
                                                  {L"insert", no_argument, 'i'},
                                                  {L"replace", no_argument, 'r'},
                                                  {L"current-buffer", no_argument, 'b'},
                                                  {L"current-job", no_argument, 'j'},
                                                  {L"current-process", no_argument, 'p'},
                                                  {L"current-selection", no_argument, 's'},
                                                  {L"current-token", no_argument, 't'},
                                                  {L"cut-at-cursor", no_argument, 'c'},
                                                  {L"function", no_argument, 'f'},
                                                  {L"tokenize", no_argument, 'o'},
                                                  {L"help", no_argument, 'h'},
                                                  {L"input", required_argument, 'I'},
                                                  {L"cursor", no_argument, 'C'},
                                                  {L"selection-start", no_argument, 'B'},
                                                  {L"selection-end", no_argument, 'E'},
                                                  {L"line", no_argument, 'L'},
                                                  {L"search-mode", no_argument, 'S'},
                                                  {L"paging-mode", no_argument, 'P'},
                                                  {L"paging-full-mode", no_argument, 'F'},
                                                  {L"is-valid", no_argument, 1},
                                                  {}};

    int opt;
    wgetopter_t w;
    while ((opt = w.wgetopt_long(argc, argv, short_options, long_options, nullptr)) != -1) {
        switch (opt) {
            case L'a': {
                append_mode = APPEND_MODE;
                break;
            }
            case L'b': {
                buffer_part = STRING_MODE;
                break;
            }
            case L'i': {
                append_mode = INSERT_MODE;
                break;
            }
            case L'r': {
                append_mode = REPLACE_MODE;
                break;
            }
            case 'c': {
                cut_at_cursor = true;
                break;
            }
            case 't': {
                buffer_part = TOKEN_MODE;
                break;
            }
            case 'j': {
                buffer_part = JOB_MODE;
                break;
            }
            case 'p': {
                buffer_part = PROCESS_MODE;
                break;
            }
            case 'f': {
                function_mode = true;
                break;
            }
            case 'o': {
                tokenize = true;
                break;
            }
            case 'I': {
                // A historical, undocumented feature. TODO: consider removing this.
                override_buffer = w.woptarg;
                break;
            }
            case 'C': {
                cursor_mode = true;
                break;
            }
            case 'B': {
                selection_start_mode = true;
                break;
            }
            case 'E': {
                selection_end_mode = true;
                break;
            }
            case 'L': {
                line_mode = true;
                break;
            }
            case 'S': {
                search_mode = true;
                break;
            }
            case 's': {
                selection_mode = true;
                break;
            }
            case 'P': {
                paging_mode = true;
                break;
            }
            case 'F': {
                paging_full_mode = true;
                break;
            }
            case 1: {
                is_valid = true;
                break;
            }
            case 'h': {
                builtin_print_help(parser, streams, cmd);
                return STATUS_CMD_OK;
            }
            case ':': {
                builtin_missing_argument(parser, streams, cmd, argv[w.woptind - 1], true);
                return STATUS_INVALID_ARGS;
            }
            case L'?': {
                builtin_unknown_option(parser, streams, cmd, argv[w.woptind - 1], true);
                return STATUS_INVALID_ARGS;
            }
            default: {
                DIE("unexpected retval from wgetopt_long");
            }
        }
    }

    if (function_mode) {
        int i;

        // Check for invalid switch combinations.
        if (buffer_part || cut_at_cursor || append_mode || tokenize || cursor_mode || line_mode ||
            search_mode || paging_mode || selection_start_mode || selection_end_mode) {
            streams.err()->append(format_string(BUILTIN_ERR_COMBO, argv[0]));
            builtin_print_error_trailer(parser, *streams.err(), cmd);
            return STATUS_INVALID_ARGS;
        }

        if (argc == w.woptind) {
            builtin_missing_argument(parser, streams, cmd, argv[0], true);
            return STATUS_INVALID_ARGS;
        }

        using rl = readline_cmd_t;
        for (i = w.woptind; i < argc; i++) {
            if (auto mc = input_function_get_code(argv[i])) {
                // Don't enqueue a repaint if we're currently in the middle of one,
                // because that's an infinite loop.
                if (mc == rl::repaint_mode || mc == rl::force_repaint || mc == rl::repaint) {
                    if (ld.is_repaint()) continue;
                }

                // HACK: Execute these right here and now so they can affect any insertions/changes
                // made via bindings. The correct solution is to change all `commandline`
                // insert/replace operations into readline functions with associated data, so that
                // all queued `commandline` operations - including buffer modifications - are
                // executed in order
                if (mc == rl::begin_undo_group || mc == rl::end_undo_group) {
                    reader_handle_command(*mc);
                } else {
                    // Inserts the readline function at the back of the queue.
                    reader_queue_ch(*mc);
                }
            } else {
                streams.err()->append(
                    format_string(_(L"%ls: Unknown input function '%ls'"), cmd, argv[i]));
                builtin_print_error_trailer(parser, *streams.err(), cmd);
                return STATUS_INVALID_ARGS;
            }
        }

        return STATUS_CMD_OK;
    }

    if (selection_mode) {
        if (rstate.selection) {
            streams.out()->append(
                {rstate.text.c_str() + rstate.selection->start, rstate.selection->length});
        }
        return STATUS_CMD_OK;
    }

    // Check for invalid switch combinations.
    if ((selection_start_mode || selection_end_mode) && (argc - w.woptind)) {
        streams.err()->append(format_string(BUILTIN_ERR_TOO_MANY_ARGUMENTS, argv[0]));
        builtin_print_error_trailer(parser, *streams.err(), cmd);
        return STATUS_INVALID_ARGS;
    }

    if ((search_mode || line_mode || cursor_mode || paging_mode) && (argc - w.woptind > 1)) {
        streams.err()->append(format_string(BUILTIN_ERR_TOO_MANY_ARGUMENTS, argv[0]));
        builtin_print_error_trailer(parser, *streams.err(), cmd);
        return STATUS_INVALID_ARGS;
    }

    if ((buffer_part || tokenize || cut_at_cursor) &&
        (cursor_mode || line_mode || search_mode || paging_mode || paging_full_mode) &&
        // Special case - we allow to get/set cursor position relative to the process/job/token.
        !(buffer_part && cursor_mode)) {
        streams.err()->append(format_string(BUILTIN_ERR_COMBO, argv[0]));
        builtin_print_error_trailer(parser, *streams.err(), cmd);
        return STATUS_INVALID_ARGS;
    }

    if ((tokenize || cut_at_cursor) && (argc - w.woptind)) {
        streams.err()->append(format_string(
            BUILTIN_ERR_COMBO2, cmd,
            L"--cut-at-cursor and --tokenize can not be used when setting the commandline"));
        builtin_print_error_trailer(parser, *streams.err(), cmd);
        return STATUS_INVALID_ARGS;
    }

    if (append_mode && !(argc - w.woptind)) {
        // No tokens in insert mode just means we do nothing.
        return STATUS_CMD_ERROR;
    }

    // Set default modes.
    if (!append_mode) {
        append_mode = REPLACE_MODE;
    }

    if (!buffer_part) {
        buffer_part = STRING_MODE;
    }

    if (line_mode) {
        streams.out()->append(
            format_string(L"%d\n", parse_util_lineno(rstate.text, rstate.cursor_pos)));
        return STATUS_CMD_OK;
    }

    if (search_mode) {
        return commandline_get_state().search_mode ? 0 : 1;
    }

    if (paging_mode) {
        return commandline_get_state().pager_mode ? 0 : 1;
    }

    if (paging_full_mode) {
        auto state = commandline_get_state();
        return (state.pager_mode && state.pager_fully_disclosed) ? 0 : 1;
    }

    if (selection_start_mode) {
        if (!rstate.selection) {
            return STATUS_CMD_ERROR;
        }
        source_offset_t start = rstate.selection->start;
        streams.out()->append(format_string(L"%lu\n", static_cast<unsigned long>(start)));
        return STATUS_CMD_OK;
    }

    if (selection_end_mode) {
        if (!rstate.selection) {
            return STATUS_CMD_ERROR;
        }
        source_offset_t end = rstate.selection->end();
        streams.out()->append(format_string(L"%lu\n", static_cast<unsigned long>(end)));
        return STATUS_CMD_OK;
    }

    // At this point we have (nearly) exhausted the options which always operate on the true command
    // line. Now we respect the possibility of a transient command line due to evaluating a wrapped
    // completion. Don't do this in cursor_mode: it makes no sense to move the cursor based on a
    // transient commandline.
    const wchar_t *current_buffer = nullptr;
    size_t current_cursor_pos{0};
    wcstring transient;
    if (override_buffer) {
        current_buffer = override_buffer;
        current_cursor_pos = std::wcslen(current_buffer);
    } else if (!ld.transient_commandlines_empty() && !cursor_mode) {
        transient = *ld.transient_commandlines_back();
        current_buffer = transient.c_str();
        current_cursor_pos = transient.size();
    } else if (rstate.initialized) {
        current_buffer = rstate.text.c_str();
        current_cursor_pos = rstate.cursor_pos;
    } else {
        // There is no command line, either because we are not interactive, or because we are
        // interactive and are still reading init files (in which case we silently ignore this).
        if (!is_interactive_session()) {
            streams.err()->append(cmd);
            streams.err()->append(L": Can not set commandline in non-interactive mode\n");
            builtin_print_error_trailer(parser, *streams.err(), cmd);
        }
        return STATUS_CMD_ERROR;
    }

    if (is_valid) {
        if (!*current_buffer) return 1;
        parser_test_error_bits_t res = parse_util_detect_errors(
            current_buffer, nullptr, true /* accept incomplete so we can tell the difference */);
        if (res & PARSER_TEST_INCOMPLETE) {
            return 2;
        }
        return res & PARSER_TEST_ERROR ? STATUS_CMD_ERROR : STATUS_CMD_OK;
    }

    switch (buffer_part) {
        case STRING_MODE: {
            begin = current_buffer;
            end = begin + std::wcslen(begin);
            break;
        }
        case PROCESS_MODE: {
            parse_util_process_extent(current_buffer, current_cursor_pos, &begin, &end, nullptr);
            break;
        }
        case JOB_MODE: {
            parse_util_job_extent(current_buffer, current_cursor_pos, &begin, &end);
            break;
        }
        case TOKEN_MODE: {
            parse_util_token_extent(current_buffer, current_cursor_pos, &begin, &end, nullptr,
                                    nullptr);
            break;
        }
        default: {
            DIE("unexpected buffer_part");
        }
    }

    if (cursor_mode) {
        if (argc - w.woptind) {
            long new_pos = fish_wcstol(argv[w.woptind]) + (begin - current_buffer);
            if (errno) {
                streams.err()->append(format_string(BUILTIN_ERR_NOT_NUMBER, cmd, argv[w.woptind]));
                builtin_print_error_trailer(parser, *streams.err(), cmd);
            }

            new_pos =
                std::max(0L, std::min(new_pos, static_cast<long>(std::wcslen(current_buffer))));
            commandline_set_buffer(current_buffer, static_cast<size_t>(new_pos));
        } else {
            size_t pos = current_cursor_pos - (begin - current_buffer);
            streams.out()->append(format_string(L"%lu\n", static_cast<unsigned long>(pos)));
        }
        return STATUS_CMD_OK;
    }

    int arg_count = argc - w.woptind;
    if (arg_count == 0) {
        write_part(begin, end, cut_at_cursor, tokenize, current_buffer, current_cursor_pos,
                   streams);
    } else if (arg_count == 1) {
        replace_part(begin, end, argv[w.woptind], append_mode, current_buffer, current_cursor_pos);
    } else {
        wcstring sb = argv[w.woptind];
        for (int i = w.woptind + 1; i < argc; i++) {
            sb.push_back(L'\n');
            sb.append(argv[i]);
        }
        replace_part(begin, end, sb.c_str(), append_mode, current_buffer, current_cursor_pos);
    }

    return STATUS_CMD_OK;
}

File: fish-rust/src/builtins/commandline.rs
use super::prelude::*;
use crate::common::{unescape_string, UnescapeFlags, UnescapeStringStyle};
use crate::input::input_function_get_code;
use crate::input_common::{CharEvent, ReadlineCmd};
use crate::parse_constants::ParserTestErrorBits;
use crate::parse_util::{
    parse_util_detect_errors, parse_util_job_extent, parse_util_lineno, parse_util_process_extent,
    parse_util_token_extent,
};
use crate::proc::is_interactive_session;
use crate::reader::{
    commandline_get_state, commandline_set_buffer, reader_handle_command, reader_queue_ch,
};
use crate::tokenizer::TokenType;
use crate::tokenizer::Tokenizer;
use crate::tokenizer::TOK_ACCEPT_UNFINISHED;
use crate::wchar::prelude::*;
use crate::wcstringutil::join_strings;
use crate::wgetopt::{wgetopter_t, wopt, woption, woption_argument_t};
use std::ops::Range;

/// Which part of the comandbuffer are we operating on.
enum TextScope {
    String,
    Job,
    Process,
    Token,
}

/// For text insertion, how should it be done.
enum AppendMode {
    // replace current text
    Replace,
    // insert at cursor position
    Insert,
    // insert at end of current token/command/buffer
    Append,
}

/// Replace/append/insert the selection with/at/after the specified string.
///
/// \param begin beginning of selection
/// \param end end of selection
/// \param insert the string to insert
/// \param append_mode can be one of REPLACE_MODE, INSERT_MODE or APPEND_MODE, affects the way the
/// test update is performed
/// \param buff the original command line buffer
/// \param cursor_pos the position of the cursor in the command line
fn replace_part(
    range: Range<usize>,
    insert: &wstr,
    insert_mode: AppendMode,
    buff: &wstr,
    cursor_pos: usize,
) {
    let mut out_pos = cursor_pos;

    let mut out = buff[..range.start].to_owned();

    match insert_mode {
        AppendMode::Replace => {
            out.push_utfstr(insert);
            out_pos = out.len();
        }
        AppendMode::Append => {
            out.push_utfstr(&buff[range.clone()]);
            out.push_utfstr(insert);
        }
        AppendMode::Insert => {
            let cursor = cursor_pos - range.start;
            assert!(range.start <= cursor);
            out.push_utfstr(&buff[range.start..cursor]);
            out.push_utfstr(&insert);
            out.push_utfstr(&buff[cursor..range.end]);
            out_pos += insert.len();
        }
    }

    out.push_utfstr(&buff[range.end..]);
    commandline_set_buffer(out, Some(out_pos));
}

/// Output the specified selection.
///
/// \param begin start of selection
/// \param end  end of selection
/// \param cut_at_cursor whether printing should stop at the surrent cursor position
/// \param tokenize whether the string should be tokenized, printing one string token on every line
/// and skipping non-string tokens
/// \param buffer the original command line buffer
/// \param cursor_pos the position of the cursor in the command line
fn write_part(
    range: Range<usize>,
    cut_at_cursor: bool,
    tokenize: bool,
    buffer: &wstr,
    cursor_pos: usize,
    streams: &mut IoStreams,
) {
    let pos = cursor_pos - range.start;

    if tokenize {
        let mut out = WString::new();
        let buff = &buffer[range];
        let mut tok = Tokenizer::new(buff, TOK_ACCEPT_UNFINISHED);
        while let Some(token) = tok.next() {
            if cut_at_cursor && token.end() >= pos {
                break;
            }

            if token.type_ == TokenType::string {
                let tmp = tok.text_of(&token);
                let unescaped =
                    unescape_string(tmp, UnescapeStringStyle::Script(UnescapeFlags::INCOMPLETE))
                        .unwrap();
                out.push_utfstr(&unescaped);
                out.push('\n');
            }
        }

        streams.out.append(out);
    } else {
        if cut_at_cursor {
            streams.out.append(&buffer[range.start..range.start + pos]);
        } else {
            streams.out.append(&buffer[range]);
        }
        streams.out.push('\n');
    }
}

/// The commandline builtin. It is used for specifying a new value for the commandline.
pub fn commandline(parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> Option<c_int> {
    let rstate = commandline_get_state();

    let mut buffer_part = None;
    let mut cut_at_cursor = false;
    let mut append_mode = None;

    let mut function_mode = false;
    let mut selection_mode = false;

    let mut tokenize = false;

    let mut cursor_mode = false;
    let mut selection_start_mode = false;
    let mut selection_end_mode = false;
    let mut line_mode = false;
    let mut search_mode = false;
    let mut paging_mode = false;
    let mut paging_full_mode = false;
    let mut is_valid = false;

    let mut range = 0..0;
    let mut override_buffer = None;

    let ld = parser.libdata();

    const short_options: &wstr = L!(":abijpctforhI:CBELSsP");
    let long_options: &[woption] = &[
        wopt(L!("append"), woption_argument_t::no_argument, 'a'),
        wopt(L!("insert"), woption_argument_t::no_argument, 'i'),
        wopt(L!("replace"), woption_argument_t::no_argument, 'r'),
        wopt(L!("current-buffer"), woption_argument_t::no_argument, 'b'),
        wopt(L!("current-job"), woption_argument_t::no_argument, 'j'),
        wopt(L!("current-process"), woption_argument_t::no_argument, 'p'),
        wopt(
            L!("current-selection"),
            woption_argument_t::no_argument,
            's',
        ),
        wopt(L!("current-token"), woption_argument_t::no_argument, 't'),
        wopt(L!("cut-at-cursor"), woption_argument_t::no_argument, 'c'),
        wopt(L!("function"), woption_argument_t::no_argument, 'f'),
        wopt(L!("tokenize"), woption_argument_t::no_argument, 'o'),
        wopt(L!("help"), woption_argument_t::no_argument, 'h'),
        wopt(L!("input"), woption_argument_t::required_argument, 'I'),
        wopt(L!("cursor"), woption_argument_t::no_argument, 'C'),
        wopt(L!("selection-start"), woption_argument_t::no_argument, 'B'),
        wopt(L!("selection-end"), woption_argument_t::no_argument, 'E'),
        wopt(L!("line"), woption_argument_t::no_argument, 'L'),
        wopt(L!("search-mode"), woption_argument_t::no_argument, 'S'),
        wopt(L!("paging-mode"), woption_argument_t::no_argument, 'P'),
        wopt(L!("paging-full-mode"), woption_argument_t::no_argument, 'F'),
        wopt(L!("is-valid"), woption_argument_t::no_argument, '\x01'),
    ];

    let mut w = wgetopter_t::new(short_options, long_options, args);
    let cmd = w.argv[0];
    while let Some(c) = w.wgetopt_long() {
        match c {
            'a' => append_mode = Some(AppendMode::Append),
            'b' => buffer_part = Some(TextScope::String),
            'i' => append_mode = Some(AppendMode::Insert),
            'r' => append_mode = Some(AppendMode::Replace),
            'c' => cut_at_cursor = true,
            't' => buffer_part = Some(TextScope::Token),
            'j' => buffer_part = Some(TextScope::Job),
            'p' => buffer_part = Some(TextScope::Process),
            'f' => function_mode = true,
            'o' => tokenize = true,
            'I' => {
                // A historical, undocumented feature. TODO: consider removing this.
                override_buffer = Some(w.woptarg.unwrap().to_owned());
            }
            'C' => cursor_mode = true,
            'B' => selection_start_mode = true,
            'E' => selection_end_mode = true,
            'L' => line_mode = true,
            'S' => search_mode = true,
            's' => selection_mode = true,
            'P' => paging_mode = true,
            'F' => paging_full_mode = true,
            '\x01' => is_valid = true,
            'h' => {
                builtin_print_help(parser, streams, cmd);
                return STATUS_CMD_OK;
            }
            ':' => {
                builtin_missing_argument(parser, streams, cmd, w.argv[w.woptind - 1], true);
                return STATUS_INVALID_ARGS;
            }
            '?' => {
                builtin_unknown_option(parser, streams, cmd, w.argv[w.woptind - 1], true);
                return STATUS_INVALID_ARGS;
            }
            _ => panic!(),
        }
    }

    let positional_args = w.argv.len() - w.woptind;

    if function_mode {
        // Check for invalid switch combinations.
        if buffer_part.is_some()
            || cut_at_cursor
            || append_mode.is_some()
            || tokenize
            || cursor_mode
            || line_mode
            || search_mode
            || paging_mode
            || selection_start_mode
            || selection_end_mode
        {
            streams.err.append(wgettext_fmt!(BUILTIN_ERR_COMBO, cmd));
            builtin_print_error_trailer(parser, streams.err, cmd);
            return STATUS_INVALID_ARGS;
        }

        if positional_args == 0 {
            builtin_missing_argument(parser, streams, cmd, L!("--function"), true);
            return STATUS_INVALID_ARGS;
        }

        type rl = ReadlineCmd;
        for arg in &w.argv[w.woptind..] {
            let Some(cmd) = input_function_get_code(arg) else {
                streams
                    .err
                    .append(wgettext_fmt!("%ls: Unknown input function '%ls'", cmd, arg));
                builtin_print_error_trailer(parser, streams.err, cmd);
                return STATUS_INVALID_ARGS;
            };
            // Don't enqueue a repaint if we're currently in the middle of one,
            // because that's an infinite loop.
            if matches!(cmd, rl::RepaintMode | rl::ForceRepaint | rl::Repaint) {
                if ld.pods.is_repaint {
                    continue;
                }
            }

            // HACK: Execute these right here and now so they can affect any insertions/changes
            // made via bindings. The correct solution is to change all `commandline`
            // insert/replace operations into readline functions with associated data, so that
            // all queued `commandline` operations - including buffer modifications - are
            // executed in order
            match cmd {
                rl::BeginUndoGroup | rl::EndUndoGroup => reader_handle_command(cmd),
                _ => {
                    // Inserts the readline function at the back of the queue.
                    reader_queue_ch(CharEvent::from_readline(cmd));
                }
            }
        }

        return STATUS_CMD_OK;
    }

    if selection_mode {
        if let Some(selection) = rstate.selection {
            streams.out.append(&rstate.text[selection]);
        }
        return STATUS_CMD_OK;
    }

    // Check for invalid switch combinations.
    if (selection_start_mode || selection_end_mode) && positional_args != 0 {
        streams
            .err
            .append(wgettext_fmt!(BUILTIN_ERR_TOO_MANY_ARGUMENTS, cmd));
        builtin_print_error_trailer(parser, streams.err, cmd);
        return STATUS_INVALID_ARGS;
    }

    if (search_mode || line_mode || cursor_mode || paging_mode) && positional_args > 1 {
        streams
            .err
            .append(wgettext_fmt!(BUILTIN_ERR_TOO_MANY_ARGUMENTS, cmd));
        builtin_print_error_trailer(parser, streams.err, cmd);
        return STATUS_INVALID_ARGS;
    }

    if (buffer_part.is_some() || tokenize || cut_at_cursor)
        && (cursor_mode || line_mode || search_mode || paging_mode || paging_full_mode)
        // Special case - we allow to get/set cursor position relative to the process/job/token.
        && (buffer_part.is_none() || !cursor_mode)
    {
        streams.err.append(wgettext_fmt!(BUILTIN_ERR_COMBO, cmd));
        builtin_print_error_trailer(parser, streams.err, cmd);
        return STATUS_INVALID_ARGS;
    }

    if (tokenize || cut_at_cursor) && positional_args != 0 {
        streams.err.append(wgettext_fmt!(
            BUILTIN_ERR_COMBO2,
            cmd,
            "--cut-at-cursor and --tokenize can not be used when setting the commandline"
        ));
        builtin_print_error_trailer(parser, streams.err, cmd);
        return STATUS_INVALID_ARGS;
    }

    if append_mode.is_some() && positional_args == 0 {
        // No tokens in insert mode just means we do nothing.
        return STATUS_CMD_ERROR;
    }

    // Set default modes.
    let append_mode = append_mode.unwrap_or(AppendMode::Replace);

    let buffer_part = buffer_part.unwrap_or(TextScope::String);

    if line_mode {
        streams.out.append(sprintf!(
            "%d\n",
            parse_util_lineno(&rstate.text, rstate.cursor_pos)
        ));
        return STATUS_CMD_OK;
    }

    if search_mode {
        return if commandline_get_state().search_mode {
            STATUS_CMD_OK
        } else {
            STATUS_CMD_ERROR
        };
    }

    if paging_mode {
        return if commandline_get_state().pager_mode {
            STATUS_CMD_OK
        } else {
            STATUS_CMD_ERROR
        };
    }

    if paging_full_mode {
        let state = commandline_get_state();
        return if state.pager_mode && state.pager_fully_disclosed {
            STATUS_CMD_OK
        } else {
            STATUS_CMD_ERROR
        };
    }

    if selection_start_mode {
        let Some(selection) = rstate.selection else {
            return STATUS_CMD_ERROR;
        };
        streams.out.append(sprintf!("%lu\n", selection.start));
        return STATUS_CMD_OK;
    }

    if selection_end_mode {
        let Some(selection) = rstate.selection else {
            return STATUS_CMD_ERROR;
        };
        streams.out.append(sprintf!("%lu\n", selection.end));
        return STATUS_CMD_OK;
    }

    // At this point we have (nearly) exhausted the options which always operate on the true command
    // line. Now we respect the possibility of a transient command line due to evaluating a wrapped
    // completion. Don't do this in cursor_mode: it makes no sense to move the cursor based on a
    // transient commandline.
    let current_buffer;
    let current_cursor_pos;
    let transient;
    if let Some(override_buffer) = &override_buffer {
        current_buffer = override_buffer;
        current_cursor_pos = current_buffer.len();
    } else if !ld.transient_commandlines.is_empty() && !cursor_mode {
        transient = ld.transient_commandlines.last().unwrap().clone();
        current_buffer = &transient;
        current_cursor_pos = transient.len();
    } else if rstate.initialized {
        current_buffer = &rstate.text;
        current_cursor_pos = rstate.cursor_pos;
    } else {
        // There is no command line, either because we are not interactive, or because we are
        // interactive and are still reading init files (in which case we silently ignore this).
        if !is_interactive_session() {
            streams.err.append(cmd);
            streams
                .err
                .append(L!(": Can not set commandline in non-interactive mode\n"));
            builtin_print_error_trailer(parser, streams.err, cmd);
        }
        return STATUS_CMD_ERROR;
    }

    if is_valid {
        if current_buffer.is_empty() {
            return Some(1);
        }
        let res = parse_util_detect_errors(current_buffer, None, /*accept_incomplete=*/ true);
        return match res {
            Ok(()) => STATUS_CMD_OK,
            Err(err) => {
                if err.contains(ParserTestErrorBits::INCOMPLETE) {
                    Some(2)
                } else {
                    STATUS_CMD_ERROR
                }
            }
        };
    }

    match buffer_part {
        TextScope::String => {
            range = 0..current_buffer.len();
        }
        TextScope::Job => {
            range = parse_util_job_extent(current_buffer, current_cursor_pos, None);
        }
        TextScope::Process => {
            range = parse_util_process_extent(current_buffer, current_cursor_pos, None);
        }
        TextScope::Token => {
            parse_util_token_extent(current_buffer, current_cursor_pos, &mut range, None);
        }
    }

    if cursor_mode {
        if positional_args != 0 {
            let arg = w.argv[w.woptind];
            let new_pos = match fish_wcstol(&arg[range.start..]) {
                Err(_) => {
                    streams
                        .err
                        .append(wgettext_fmt!(BUILTIN_ERR_NOT_NUMBER, cmd, arg));
                    builtin_print_error_trailer(parser, streams.err, cmd);
                    0
                }
                Ok(num) => num,
            };

            let new_pos = std::cmp::min(new_pos.max(0) as usize, current_buffer.len());
            commandline_set_buffer(current_buffer.to_owned(), Some(new_pos));
        } else {
            streams.out.append(sprintf!("%lu\n", current_cursor_pos));
        }
        return STATUS_CMD_OK;
    }

    if positional_args == 0 {
        write_part(
            range,
            cut_at_cursor,
            tokenize,
            current_buffer,
            current_cursor_pos,
            streams,
        );
    } else if positional_args == 1 {
        replace_part(
            range,
            args[w.woptind],
            append_mode,
            current_buffer,
            current_cursor_pos,
        );
    } else {
        let sb = join_strings(&w.argv[w.woptind..], '\n');
        replace_part(range, &sb, append_mode, current_buffer, current_cursor_pos);
    }

    STATUS_CMD_OK
}
